File: Sources/CodexManager/Behavior/AccountsTransferMerge.swift
import Foundation

enum AccountsTransferMerge {
    static func applying(
        importedAccounts: [StoredAccount],
        selectedAccountIDs: Set<String>,
        to store: AccountsStore,
        now: Int64
    ) -> (store: AccountsStore, result: AccountsImportResult) {
        var merged = store
        var insertedCount = 0
        var updatedCount = 0

        for account in importedAccounts where selectedAccountIDs.contains(account.id) {
            var incoming = account
            incoming.updatedAt = now
            if incoming.addedAt <= 0 {
                incoming.addedAt = now
            }

            if let existingIndex = merged.accounts.firstIndex(where: { AccountIdentity.matches($0, incoming) }) {
                let existing = merged.accounts[existingIndex]
                incoming.id = existing.id
                incoming.addedAt = existing.addedAt
                merged.accounts[existingIndex] = incoming
                updatedCount += 1
                continue
            }

            if merged.accounts.contains(where: { $0.id == incoming.id }) {
                incoming.id = UUID().uuidString
            }
            merged.accounts.append(incoming)
            insertedCount += 1
        }

        return (
            merged,
            AccountsImportResult(insertedCount: insertedCount, updatedCount: updatedCount)
        )
    }
}

File: apps/windows/src/shared/domain/accounts-transfer-merge.ts
import type { AccountsImportResult } from "../models/account-transfer";
import type { AccountsStore, StoredAccount } from "../models/accounts";
import { storedAccountsMatch } from "./account-identity";

export interface AccountsTransferMergeResult {
  store: AccountsStore;
  result: AccountsImportResult;
}

export interface AccountsTransferMergeOptions {
  idGenerator?: () => string;
}

export function applyAccountsTransferMerge(
  importedAccounts: readonly StoredAccount[],
  selectedAccountIds: ReadonlySet<string>,
  store: AccountsStore,
  now: number,
  options: AccountsTransferMergeOptions = {}
): AccountsTransferMergeResult {
  const idGenerator = options.idGenerator ?? (() => globalThis.crypto.randomUUID());
  const merged: AccountsStore = {
    ...store,
    accounts: [...store.accounts]
  };
  let insertedCount = 0;
  let updatedCount = 0;

  for (const account of importedAccounts) {
    if (!selectedAccountIds.has(account.id)) {
      continue;
    }

    const incoming: StoredAccount = {
      ...account,
      updatedAt: now,
      addedAt: account.addedAt > 0 ? account.addedAt : now
    };

    const existingIndex = merged.accounts.findIndex((existing) => storedAccountsMatch(existing, incoming));
    if (existingIndex >= 0) {
      const existing = merged.accounts[existingIndex];
      if (!existing) {
        continue;
      }

      merged.accounts[existingIndex] = {
        ...incoming,
        id: existing.id,
        addedAt: existing.addedAt
      };
      updatedCount += 1;
      continue;
    }

    if (merged.accounts.some((existing) => existing.id === incoming.id)) {
      incoming.id = idGenerator();
    }

    merged.accounts.push(incoming);
    insertedCount += 1;
  }

  return {
    store: merged,
    result: {
      insertedCount,
      updatedCount
    }
  };
}
